### PR TITLE
Fix for UnivariatePolynomial::__eq__

### PR DESCRIPTION
**Pull request by myluszczak**
The current implementation of `UnivariatePolynomial::__eq__` checks if the dictionaries of the two polynomials are exactly equal, and will return false if comparing polynomials with the dictionaries `{{0,0},{1,2}}` and `{{1,2}}` representing `0 + 2x` and `2x` respectively.  Since addition does not remove the entries of a dictionary corresponding to terms that have zero coefficients, this can lead to a situation where SymEngine tells us that `(x+1) + (x-1) != 2x`, as shown below.
[test.zip](https://github.com/symengine/symengine/files/159562/test.zip)

My implementation replaces the call to `map_uint_mpz_eq` with a call to `poly_dict_map_uint_mpz_eq`, which returns true if two dictionaries are the same (excluding those entries where the coefficient is zero.)

### CODE REVIEW

**Comment by certik** [comment]
I wonder if we should just make sure when the polynomial is created, that it is always in a canonical form, so that then you can literally just compare the dictionaries?

@isuruf what do you think?

**Comment by isuruf** [comment]
Yes, that should be easier.
We can use `UnivariatePolynomial::from_dict` to do that. Current implementation is not necessary which checks if the dictionary size is 0 or 1, which is not needed. `UnivariatePolynomial::from_dict` should always return a polynomial.

**Comment by myluszczak** [comment]
@isuruf So, something like this?

If we do it this way, we should probably add code to is_canonical to check that the dictionary contains no terms with zero coefficients.  This would make the canonical form of the polynomial "0" be a polynomial with an empty dictionary, which would require changing some code in the tester and some of the code in printer.cpp.

**Comment by isuruf** [comment]
@myluszczak, exactly. I think this looks good.

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -30,7 +30,12 @@ bool UnivariatePolynomial::is_canonical(const unsigned int &degree_, const map_u
     unsigned int prev_degree = (--ordered.end())->first;
     if (prev_degree != degree_)
         return false;
-
+    //Check if dictionary contains terms with coeffienct 0
+    /*for(auto itter = dict.begin(); itter != dict.end(); itter++){
+        if(0 == itter->second)
+	    return false;
+    }*/
```
Comment (isuruf) [inline symengine/polynomial.cpp]: You can uncomment this code

**Comment by isuruf** [comment]
There are currently 3 ways of creating a polynomial.

1. `UnivariatePolynomial(const RCP<const Symbol> &var, const unsigned int &degree, map_uint_mpz&& dict);`
2. `UnivariatePolynomial(const RCP<const Symbol> &var, const std::vector<integer_class> &v);`
3. `static RCP<const UnivariatePolynomial> from_dict(const RCP<const Symbol> &var, map_uint_mpz &&d);`

Instead of that we can do,

1. `UnivariatePolynomial(const RCP<const Symbol> &var, const unsigned int &degree, map_uint_mpz&& dict);`
2. `static RCP<const UnivariatePolynomial> from_vec(const RCP<const Symbol> &var, const std::vector<integer_class> &v);`
3. `static RCP<const UnivariatePolynomial> from_dict(const RCP<const Symbol> &var, map_uint_mpz &&d);`

Only method 2 and 3 will call 1 and all other references to 1 will be replaced by 3. 
What do you think?

**Comment by myluszczak** [comment]
@isuruf I think that's a good plan.  

Thanks for the head's up on the redundant checks in get_args; I missed that when I copied the code form the original implementation of from_dict.

**Comment by myluszczak** [comment]
In the above commit I've done what I though would implement isuruf's idea above, however on my machine compilation stops when linking test_arit with the error message
`[ 42%] Linking CXX executable test_arit
../../libsymengine.a(polynomial.cpp.o): In function `SymEngine::mul_uni_poly(Teuchos::RCP<SymEngine::UnivariatePolynomial const>, Teuchos::RCP<SymEngine::UnivariatePoly                                       nomial const>)':
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/polynomial.cpp:290: undefined reference to `SymEngine::UnivariatePolynomial::from_vec(Teuchos::RCP<SymEngine::Symbol                                        const> const&, std::vector<__gmp_expr<__mpz_struct [1], __mpz_struct [1]>, std::allocator<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > const&)'
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/polynomial.cpp:292: undefined reference to `SymEngine::UnivariatePolynomial::from_vec(Teuchos::RCP<SymEngine::Symbol                                        const> const&, std::vector<__gmp_expr<__mpz_struct [1], __mpz_struct [1]>, std::allocator<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > const&)'
../../libsymengine.a(series_generic.cpp.o): In function `SymEngine::UnivariatePolynomial::create(Teuchos::RCP<SymEngine::Symbol const> const&, std::vector<__gmp_expr<__                                       mpz_struct [1], __mpz_struct [1]>, std::allocator<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > const&)':
/home/myluszcz/ECS193/SYMENGINE/symengine/symengine/polynomial.h:49: undefined reference to `SymEngine::UnivariatePolynomial::from_vec(Teuchos::RCP<SymEngine::Symbol co                                       nst> const&, std::vector<__gmp_expr<__mpz_struct [1], __mpz_struct [1]>, std::allocator<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > const&)'
collect2: error: ld returned 1 exit status
symengine/tests/basic/CMakeFiles/test_arit.dir/build.make:104: recipe for target 'symengine/tests/basic/test_arit' failed
make[2]: *** [symengine/tests/basic/test_arit] Error 1
CMakeFiles/Makefile2:381: recipe for target 'symengine/tests/basic/CMakeFiles/test_arit.dir/all' failed
make[1]: *** [symengine/tests/basic/CMakeFiles/test_arit.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
`
Since it's late, I'm going to shower and go to bed.

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -73,30 +69,25 @@ int UnivariatePolynomial::compare(const Basic &o) const
...
-                return Mul::from_dict(integer(d.begin()->second),
-                    {{var, integer(d.begin()->first)}});
+    for(auto itter = d.begin(); itter != d.end(); itter++){
+        if(0_z == itter->second)
+  	    d.erase(itter);
+    }
+    return make_rcp<const UnivariatePolynomial>(var, (--(d.end()))->first, std::move(d));
+}
+
+RCP<const UnivariatePolynomial> from_vec(const RCP<const Symbol> &var, const std::vector<integer_class> &v){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Change the name to `UnivariatePolynomial::from_vec`

File: symengine/polynomial.cpp [hunk 3]
```diff
@@ -53,7 +49,7 @@ std::size_t UnivariatePolynomial::__hash__() const
 bool UnivariatePolynomial::__eq__(const Basic &o) const
 {
     if (eq(*var_, *(static_cast<const UnivariatePolynomial &>(o).var_)) and
-        map_uint_mpz_eq(dict_, static_cast<const UnivariatePolynomial &>(o).dict_))
+        poly_dict_map_uint_mpz_eq(dict_, static_cast<const UnivariatePolynomial &>(o).dict_))
```
Comment (isuruf) [inline symengine/polynomial.cpp]: You can now keep the old behaviour

**Comment by isuruf** [comment]
Can you rebase on top of master and fix merge conflicts? @chenchfort should be able to help you out. If not let me know if you need my help.

**Comment by shivamvats** [comment]
@myluszczak Are you developing in debug mode?

**Comment by myluszczak** [comment]
@shivamvats Not in general, no.  I just compiled and tested it in debug mode, though, and it still works on my machine.

File: symengine/polynomial.cpp [hunk 4]
```diff
@@ -73,30 +69,28 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
-            if (d.begin()->second == 0)
-                return zero;
-            else if (d.begin()->second == 1)
-                return pow(var, integer(d.begin()->first));
-            else
-                return Mul::from_dict(integer(d.begin()->second),
-                    {{var, integer(d.begin()->first)}});
+    for(auto itter = d.begin(); itter != d.end(); itter++){
+        if(integer_class(0) == itter->second)
+  	    d.erase(itter);
```
Comment (isuruf) [inline symengine/polynomial.cpp]: This is probably what is causing the segfault. If you delete while iterating, the iterator becomes invalid.

**Comment by myluszczak** [comment]
@isuruf So, a better implementation would be to create a new dictionary and only add to that if the coefficient is nonzero.

**Comment by isuruf** [comment]
See this answer, http://stackoverflow.com/a/4600592/4768820

**Comment by myluszczak** [comment]
I see, thank you.

**Comment by myluszczak** [comment]
@isuruf This commit implements your suggestion.  Is there a way to cancel the appveyor builds?  At this point two of them just time out, so right now it just clogs the queue.

**Comment by isuruf** [comment]
I've stopped the appveyor build.

File: symengine/dict.cpp [hunk 5]
```diff
@@ -182,6 +182,31 @@ int map_uint_mpz_compare(const map_uint_mpz &A, const map_uint_mpz &B)
     return 0;
 }
 
+bool poly_dict_map_uint_mpz_eq(const map_uint_mpz &a, const map_uint_mpz &b){
```
Comment (isuruf) [inline symengine/dict.cpp]: You can remove this code now

File: symengine/polynomial.cpp [hunk 6]
```diff
@@ -12,25 +12,21 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
-
-    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
-}
-
 bool UnivariateIntPolynomial::is_canonical(const unsigned int &degree_, const map_uint_mpz& dict) const
 {
     map_uint_mpz ordered(dict.begin(), dict.end());
-    unsigned int prev_degree = (--ordered.end())->first;
-    if (prev_degree != degree_)
+    if(ordered.size() != 0){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Add space after `if`

File: symengine/polynomial.cpp [hunk 7]
```diff
@@ -12,25 +12,21 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
-            dict_add_term(dict_, v[i], i);
-    }
-    degree_ = v.size() - 1;
-
-    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
-}
-
 bool UnivariateIntPolynomial::is_canonical(const unsigned int &degree_, const map_uint_mpz& dict) const
 {
     map_uint_mpz ordered(dict.begin(), dict.end());
```
Comment (isuruf) [inline symengine/polynomial.cpp]: You don't need this copy now

File: symengine/polynomial.cpp [hunk 8]
```diff
@@ -12,25 +12,21 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
-    if (prev_degree != degree_)
+    if(ordered.size() != 0){
+        unsigned int prev_degree = (--ordered.end())->first;
+        if (prev_degree != degree_)
+            return false;
+    } else if(0 != degree_)
         return false;
-
+    //Check if dictionary contains terms with coeffienct 0
+    for(auto itter = dict.begin(); itter != dict.end(); itter++){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Add space after `for`

File: symengine/polynomial.cpp [hunk 9]
```diff
@@ -12,25 +12,21 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
+        if(0 == itter->second)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Add space after `if`

**Comment by myluszczak** [comment]
@isuruf Thanks

File: symengine/polynomial.cpp [hunk 10]
```diff
@@ -73,30 +69,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
-        } else {
-            if (d.begin()->second == 0)
-                return zero;
-            else if (d.begin()->second == 1)
-                return pow(var, integer(d.begin()->first));
-            else
-                return Mul::from_dict(integer(d.begin()->second),
-                    {{var, integer(d.begin()->first)}});
+    auto itter = d.begin();
+    while(itter != d.end()){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Add space after `while` and before `{`

File: symengine/polynomial.cpp [hunk 11]
```diff
@@ -144,49 +160,46 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
+    if (dict_.size() == 1 and dict_.begin()->first == 1 and dict_.begin()->second == 1)
+            return true;
     return false;
 }
 
 bool UnivariateIntPolynomial::is_mul() const {
-    if (dict_.size() == 1 and dict_.begin()->first != 0 and
-            dict_.begin()->second != 1 and dict_.begin()->second != 0)
-        return true;
+    if (dict_.size() == 1 and dict_.begin()->first != 0 and dict_.begin()->second != 1 and dict_.begin()->second != 0)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Keep the formatting as before

File: symengine/polynomial.h [hunk 12]
```diff
@@ -27,12 +27,10 @@ class UnivariateIntPolynomial : public Basic{
     IMPLEMENT_TYPEID(UNIVARIATEINTPOLYNOMIAL)
     //! Constructor of UnivariateIntPolynomial class
     UnivariateIntPolynomial(const RCP<const Symbol> &var, const unsigned int &degree, map_uint_mpz&& dict);
-    //! Constructor using a dense vector of integer_class coefficients
-    UnivariateIntPolynomial(const RCP<const Symbol> &var, const std::vector<integer_class> &v);
 
     static RCP<const UnivariateIntPolynomial> create(const RCP<const Symbol> &var,
-            const std::vector<integer_class> &v) {
-        return make_rcp<const UnivariateIntPolynomial>(var, v);
+						  const std::vector<integer_class> &v) {
```
Comment (isuruf) [inline symengine/polynomial.h]: Keep the formatting as before.

**Comment by isuruf** [comment]
Looks good to me now. You have to do some formatting. (Things like space after `if`, etc.)
Also you need to rebase on top of master.

**Comment by myluszczak** [comment]
@isuruf Done.  I think.

File: symengine/polynomial.cpp [hunk 13]
```diff
@@ -108,9 +106,25 @@ void UnivariateIntPolynomial::dict_add_term(map_uint_mpz &d, const integer_class
...
+            else
+                args.push_back(Mul::from_dict(integer(d.begin()->second), {{var_, one}}));
+        } else {
+            if (d.begin()->second == 1)
+	        args.push_back( pow(var_, integer(d.begin()->first)));
+            else
+	        args.push_back( Mul::from_dict(integer(d.begin()->second),{{var_, integer(d.begin()->first)}}));
+        }
     }
+    if(dict_.empty())
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space after `if`

File: symengine/polynomial.cpp [hunk 14]
```diff
@@ -144,26 +158,28 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
-    if (dict_.size() == 1 and dict_.begin()->second == 0)
+    if (dict_.empty())
         return true;
     return false;
 }
 
 bool UnivariateIntPolynomial::is_one() const {
     if (dict_.size() == 1 and dict_.begin()->second == 1 and
-            dict_.begin()->first == 0)
+	    dict_.begin()->first == 0)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Keep this as it was

File: symengine/polynomial.cpp [hunk 15]
```diff
@@ -144,26 +158,28 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
         return true;
     return false;
 }
 
 bool UnivariateIntPolynomial::is_minus_one() const {
     if (dict_.size() == 1 and dict_.begin()->second == -1 and
-            dict_.begin()->first == 0)
+	    dict_.begin()->first == 0)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same here

File: symengine/polynomial.cpp [hunk 16]
```diff
@@ -171,21 +187,21 @@ bool UnivariateIntPolynomial::is_integer() const {
 
 bool UnivariateIntPolynomial::is_symbol() const {
     if (dict_.size() == 1 and dict_.begin()->first == 1 and
-            dict_.begin()->second == 1)
+	    dict_.begin()->second == 1)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same here

File: symengine/polynomial.cpp [hunk 17]
```diff
@@ -144,26 +158,28 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
...
 bool UnivariateIntPolynomial::is_minus_one() const {
     if (dict_.size() == 1 and dict_.begin()->second == -1 and
-            dict_.begin()->first == 0)
+	    dict_.begin()->first == 0)
         return true;
     return false;
 }
 
 bool UnivariateIntPolynomial::is_integer() const {
+    if(dict_.empty())
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space after `if`

File: symengine/polynomial.cpp [hunk 18]
```diff
@@ -73,30 +67,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
+    if(!d.empty())
+        degree = (--(d.end()))->first;
+    return make_rcp<const UnivariateIntPolynomial>(var, degree, std::move(d));
+}
+
+RCP<const UnivariateIntPolynomial> UnivariateIntPolynomial::from_vec(const RCP<const Symbol> &var, const std::vector<integer_class> &v){
+    map_uint_mpz dict;
+    unsigned int degree = 0;
+    for(unsigned int i = 0; i < v.size(); i++){
+        if(0 != v[i]){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same here

File: symengine/polynomial.cpp [hunk 19]
```diff
@@ -73,30 +67,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
+        if (integer_class(0) == itter->second){
+	    auto toErase = itter;
+	    itter++;
+	    d.erase(toErase);
+        }else{
+	    itter++;
+	}
+    }
+    unsigned int degree = 0;
+    if(!d.empty())
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same here

File: symengine/polynomial.cpp [hunk 20]
```diff
@@ -12,25 +12,20 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
 bool UnivariateIntPolynomial::is_canonical(const unsigned int &degree_, const map_uint_mpz& dict) const
 {
-    map_uint_mpz ordered(dict.begin(), dict.end());
-    unsigned int prev_degree = (--ordered.end())->first;
-    if (prev_degree != degree_)
+    if (dict.size() != 0){
+        unsigned int prev_degree = (--dict.end())->first;
+        if (prev_degree != degree_)
+            return false;
+    } else if(0 != degree_)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same here.
Also `{` after `)`

File: symengine/polynomial.cpp [hunk 21]
```diff
@@ -12,25 +12,20 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
...
-
-    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
-}
-
 bool UnivariateIntPolynomial::is_canonical(const unsigned int &degree_, const map_uint_mpz& dict) const
 {
-    map_uint_mpz ordered(dict.begin(), dict.end());
-    unsigned int prev_degree = (--ordered.end())->first;
-    if (prev_degree != degree_)
+    if (dict.size() != 0){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space before `{`

File: symengine/polynomial.cpp [hunk 22]
```diff
@@ -73,30 +67,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
-            else
-                return Mul::from_dict(integer(d.begin()->second),
-                    {{var, integer(d.begin()->first)}});
+    auto itter = d.begin();
+    while (itter != d.end()) {
+        if (integer_class(0) == itter->second){
+	    auto toErase = itter;
+	    itter++;
+	    d.erase(toErase);
+        }else{
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space before `{` and after `}`

File: symengine/polynomial.cpp [hunk 23]
```diff
@@ -73,30 +67,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
-            if (d.begin()->second == 0)
-                return zero;
-            else if (d.begin()->second == 1)
-                return pow(var, integer(d.begin()->first));
-            else
-                return Mul::from_dict(integer(d.begin()->second),
-                    {{var, integer(d.begin()->first)}});
+    auto itter = d.begin();
+    while (itter != d.end()) {
+        if (integer_class(0) == itter->second){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space before `{`

File: symengine/polynomial.cpp [hunk 24]
```diff
@@ -73,30 +67,34 @@ int UnivariateIntPolynomial::compare(const Basic &o) const
...
+    unsigned int degree = 0;
+    if(!d.empty())
+        degree = (--(d.end()))->first;
+    return make_rcp<const UnivariateIntPolynomial>(var, degree, std::move(d));
+}
+
+RCP<const UnivariateIntPolynomial> UnivariateIntPolynomial::from_vec(const RCP<const Symbol> &var, const std::vector<integer_class> &v){
+    map_uint_mpz dict;
+    unsigned int degree = 0;
+    for(unsigned int i = 0; i < v.size(); i++){
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Space before `{` and after `for`

File: symengine/polynomial.cpp [hunk 25]
```diff
@@ -108,9 +106,25 @@ void UnivariateIntPolynomial::dict_add_term(map_uint_mpz &d, const integer_class
 
 vec_basic UnivariateIntPolynomial::get_args() const {
     vec_basic args;
+    map_uint_mpz d;
     for (const auto &p: dict_) {
-        args.push_back(UnivariateIntPolynomial::from_dict(var_, {{p.first, p.second}}));
+        d = {{p.first, p.second}};
+        if (d.begin()->first == 0)
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Use `{` here even though the statement below is just one line

File: symengine/printer.cpp [hunk 26]
```diff
@@ -320,6 +320,8 @@ void StrPrinter::bvisit(const UnivariateIntPolynomial &x) {
         //corner cases of only first term handled successfully, switch the bool
         first = false;
     }
+    if(x.dict_.size() == 0)
```
Comment (isuruf) [inline symengine/printer.cpp]: Space after `if`

File: symengine/dict.h [hunk 27]
```diff
@@ -94,7 +94,8 @@ bool vec_basic_eq_perm(const vec_basic &a, const vec_basic &b);
 bool map_uint_mpz_eq(const map_uint_mpz &a, const map_uint_mpz &b);
 //! \return true if the two multisets `a` and `b` are equal. Otherwise false.
 bool multiset_basic_eq(const multiset_basic &a, const multiset_basic &b);
-
+// return true if the two dictionaries are equal, ignoring entries where the coeffient is zero.  Otherwise false.
+bool poly_dict_map_uint_mpz_eq(const map_uint_mpz &a, const map_uint_mpz &b);
```
Comment (isuruf) [inline symengine/dict.h]: You don't need this method anymore

**Comment by isuruf** [comment]
+1. If you can rebase on top of master, we can get this merged

**Comment by myluszczak** [comment]
I'm a bit new to git.  That'd be `git rebase master` in the `polydicteq` branch, right?

**Comment by isuruf** [comment]
Yes

**Comment by chenchfort** [comment]
You'll need to resolve the conflict.

**Comment by myluszczak** [comment]
@chenchfort wait, which conflict?

**Comment by chenchfort** [comment]
Have you rebased? Did it complain about conflicts?

**Comment by myluszczak** [comment]
I'm rebasing now, and it is complaining about conflicts currently

**Comment by myluszczak** [comment]
@chenchfort  Alright, so I finished the rebase, but now git refuses to push 

``To https://github.com/chenchfort/symengine
 ! [rejected]        polydicteq -> polydicteq (non-fast-forward)
error: failed to push some refs to 'https://github.com/chenchfort/symengine'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Comment by chenchfort** [comment]
Ok force pushed. Looks good.

**Comment by isuruf** [comment]
Closed in favour of #855

**Pull request closed without merging**